traffic_ops/traffic_ops_golang/server/status_count.go

```diff
@@ -0,0 +1,70 @@
+package server
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import (
+	"database/sql"
+	"errors"
+	"net/http"
+
+	"github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/api"
+)
+
+func GetServersStatusCountsHandler(w http.ResponseWriter, r *http.Request) {
+	inf, userErr, sysErr, errCode := api.NewInfo(r, nil, nil)
+	if userErr != nil || sysErr != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, errCode, userErr, sysErr)
+		return
+	}
+	defer inf.Close()
+
+	statusCounts, err := getServersStatusCounts(inf.Tx.Tx, inf.Params["type"])
+	if err != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, http.StatusInternalServerError, nil, errors.New("getting servers status counts: "+err.Error()))
+		return
+	}
+	api.WriteResp(w, r, statusCounts)
+}
+
+func getServersStatusCounts(tx *sql.Tx, typeName string) (map[string]int, error) {
+	q := `
+SELECT status.name, count(server.id)
+FROM server
+JOIN status ON server.status = status.id
+JOIN type ON server.type = type.id
+WHERE type.name ~ $1
```

Comment:
Why use a regex on that?

Reply:
to replicate the Perl version

Reply:
I never knew that about Perl. You wanna document that? The `type` query parameter docs don't indicate that it supports POSIX regex.

Reply:
Eh, I don't really want to document that. IMO it should really be `=`, but `~` makes it also count things like `EDGE_FOO` and `EDGE_BAR` if you pass `?type=EDGE`.

Reply:
Hm, regexes can be constructed to both be extremely CPU and Memory intensive. This will be a security vulnerability, if it's exposed with Self Service.

Since it's never been documented, what about changing it to `=`, or if you think people are using it for prefixes, maybe `LIKE '%' + $1` or `LIKE '%' + $1 + '%'`?

Reply:
There's no input restriction on Type names, they can legally include backslashes. `strings.Replace` isn't powerful enough to guard against arbitrary bad Type names, I think.

Reply:
Escaping `%` and `_` was more to prevent accidents than malice. LIKE isn't overly expensive, it's not really a security risk for someone to do their own `%` if they really want to.

Though we could prevent that with an additional `strings.Replace(i, "\", "\\")` if we wanted.

Reply:
what if I did `\%`? -> `\\%` -> `\\\\%` 

but if it's not a problem it's not a problem. I think `LIKE` is a better choice than parsing arbitrary regular expressions.

Also, if parsing user-input regexes can be considered a security risk then Content Invalidation Jobs are a huge hole.

Reply:
>what if I did \%? -> \\% -> \\\\%

The \ replace needs to come before the % replace, so `\%` becomes `\\\%`, which will be parsed by Postgres as a literal \ and a literal %.

> Also, if parsing user-input regexes can be considered a security risk then Content Invalidation Jobs are a huge hole.

Indeed. :(

Reply:
If we're going to depart from the original Perl behavior, then I'll just make it `=`.

Reply:
that sounds fine to me